### PR TITLE
Fix Explosion + Galvanize + Volt Absorb and Mind Blown

### LINKED ISSUE #1961: [battle_engine]Volt absorb stops explosion from damaging other pokémon
EDIT : [this PR](https://github.com/rh-hideout/pokeemerald-expansion/pull/1975) addresses all the problems mentionned, this issue can be closed once the PR is merged!

To be accurate, if you use explosion with the galvanize ability, in a double battle, and one of the mon has motor drive or volt absorb.
When voltabsorb/motor drive is activated, all the subsequent targets will not get damaged (and they should, did the test in showdown, plus it makes sense).
I thought I could fix it when working on https://github.com/rh-hideout/pokeemerald-expansion/pull/1958 but I can't seem to.
I think that explosion's battlescript should be updated, because it seems to use the same old version that earthquake was using; makes me think it might be the reason why surf doesn't have the same issue with water absorb/storm drain in double, or earthquake with levitate.
I tried to mess a little with the battlescripts but this process is inherently flawed, because if it luckily works at some point and I don't know why, I wouldn't know if I've broken something else.
If someone with a strong understanding of these battlescripts and the moveend command could look into this, and perhaps give me some pointers, it'd be very helpful.

### PR DESCRIPTION
**Pull request by DizzyEggg**
Fixes #1961

Replacement for an old PR #1975, but my fix is less invasive. 
I only modified(by saving a battler) and made use of a previously unused command jumpifabilitypresent.
I simplified Explosion / Mind Blown battlescript, so it doesn't loop the targets itself, but handles the job to the moveend command.
Fixed Mind Blown, because it was always fainting the user before.

### CODE REVIEW

**Comment by DizzyEggg** [comment]
Added Tests for Mind Blown and updated tests for Volt Absorb, Damp and Explosion.
Ready for review

File: test/move_effect_mind_blown.c [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+    }
+}
+
+DOUBLE_BATTLE_TEST("Mind Blown causes everyone to faint in a double battle")
+{
+    GIVEN {
+        PLAYER(SPECIES_WOBBUFFET) { HP(HP_TEST / 2) ; MaxHP(HP_TEST) ;} ;
+        PLAYER(SPECIES_WYNAUT) { HP(1); } ;
+        OPPONENT(SPECIES_ABRA) { HP(1); } ;
+        OPPONENT(SPECIES_KADABRA) { HP(1); } ;
```
Comment (mrgriffin) [inline test/move_effect_mind_blown.c:94]:
```suggestion
        PLAYER(SPECIES_WOBBUFFET) { HP(HP_TEST / 2); MaxHP(HP_TEST); }
        PLAYER(SPECIES_WYNAUT) { HP(1); }
        OPPONENT(SPECIES_ABRA) { HP(1); }
        OPPONENT(SPECIES_KADABRA) { HP(1); }
```

File: test/move_effect_mind_blown.c [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+        OPPONENT(SPECIES_KADABRA);
+    } WHEN {
+        TURN { MOVE(playerLeft, MOVE_MIND_BLOWN, criticalHit:FALSE); }
```
Comment (mrgriffin) [inline test/move_effect_mind_blown.c:97]:
```suggestion
        TURN { MOVE(playerLeft, MOVE_MIND_BLOWN, criticalHit: FALSE); }
```

File: test/move_effect_mind_blown.c [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+        ANIMATION(ANIM_TYPE_MOVE, MOVE_MIND_BLOWN, playerLeft);
+        MESSAGE("Wobbuffet fainted!");
+    }
+}
+
+SINGLE_BATTLE_TEST("Mind Blown causes the user & the target to faint when below 1/2 of its HP")
+{
+    GIVEN {
+        PLAYER(SPECIES_WOBBUFFET) { HP(HP_TEST / 2) ; MaxHP(HP_TEST) ;} ;
+        OPPONENT(SPECIES_WOBBUFFET) { HP(1); } ;
```
Comment (mrgriffin) [inline test/move_effect_mind_blown.c:75]:
```suggestion
        PLAYER(SPECIES_WOBBUFFET) { HP(HP_TEST / 2) ; MaxHP(HP_TEST); }
        OPPONENT(SPECIES_WOBBUFFET) { HP(1); }
```

File: test/move_effect_mind_blown.c [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+        HP_BAR(player, hp: 0);
+        ANIMATION(ANIM_TYPE_MOVE, MOVE_MIND_BLOWN, player);
+        MESSAGE("Wobbuffet fainted!");
+    }
+}
+
+DOUBLE_BATTLE_TEST("Mind Blown causes the user to faint when below 1/2 of its HP in a double battle")
+{
+    GIVEN {
+        PLAYER(SPECIES_WOBBUFFET) { HP(HP_TEST / 2) ; MaxHP(HP_TEST) ;} ;
```
Comment (mrgriffin) [inline test/move_effect_mind_blown.c:58]:
```suggestion
        PLAYER(SPECIES_WOBBUFFET) { HP(HP_TEST / 2); MaxHP(HP_TEST); }
```

File: test/move_effect_mind_blown.c [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+        HP_BAR(playerLeft, hp: HP_TEST / 2);
+        ANIMATION(ANIM_TYPE_MOVE, MOVE_MIND_BLOWN, playerLeft);
+        NOT MESSAGE("Wobbuffet fainted!"); // Wobb had more than 1/2 of its HP, so it can't faint.
+    }
+}
+
+SINGLE_BATTLE_TEST("Mind Blown causes the user to faint when below 1/2 of its HP")
+{
+    GIVEN {
+        PLAYER(SPECIES_WOBBUFFET) { HP(HP_TEST / 2) ; MaxHP(HP_TEST) ;} ;
```
Comment (mrgriffin) [inline test/move_effect_mind_blown.c:44]:
```suggestion
        PLAYER(SPECIES_WOBBUFFET) { HP(HP_TEST / 2); MaxHP(HP_TEST); }
```

File: test/move_effect_mind_blown.c [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+        OPPONENT(SPECIES_WOBBUFFET);
+    } WHEN {
+        TURN { MOVE(player, MOVE_MIND_BLOWN);}
```
Comment (mrgriffin) [inline test/move_effect_mind_blown.c:47]:
```suggestion
        TURN { MOVE(player, MOVE_MIND_BLOWN); }
```

File: test/move_effect_mind_blown.c [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+        HP_BAR(player, hp: HP_TEST / 2);
+        ANIMATION(ANIM_TYPE_MOVE, MOVE_MIND_BLOWN, player);
+        NOT MESSAGE("Wobbuffet fainted!"); // Wobb had more than 1/2 of its HP, so it can't faint.
+    }
+}
+
+DOUBLE_BATTLE_TEST("Mind Blown makes the user lose 1/2 of its HP in a double battle")
+{
+    GIVEN {
+        PLAYER(SPECIES_WOBBUFFET) { HP(HP_TEST) ; MaxHP(HP_TEST) ;} ;
```
Comment (mrgriffin) [inline test/move_effect_mind_blown.c:28]:
```suggestion
        PLAYER(SPECIES_WOBBUFFET) { HP(HP_TEST); MaxHP(HP_TEST); }
```

File: test/move_effect_mind_blown.c [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+{
+    ASSUME(gBattleMoves[MOVE_MIND_BLOWN].effect == EFFECT_MIND_BLOWN);
+}
+
+#define HP_TEST (400)
+
+SINGLE_BATTLE_TEST("Mind Blown makes the user lose 1/2 of its HP")
+{
+    GIVEN {
+        PLAYER(SPECIES_WOBBUFFET) { HP(HP_TEST) ; MaxHP(HP_TEST) ;} ;
```
Comment (mrgriffin) [inline test/move_effect_mind_blown.c:14]:
```suggestion
        PLAYER(SPECIES_WOBBUFFET) { HP(HP_TEST); MaxHP(HP_TEST); }
```

File: test/move_effect_explosion.c [hunk 9]
```diff
@@ -49,5 +65,29 @@ SINGLE_BATTLE_TEST("Explosion causes the user to faint even if it has no effect"
...
+    }
+}
+
+DOUBLE_BATTLE_TEST("Explosion causes everyone to faint in a double battle")
+{
+    GIVEN {
+        PLAYER(SPECIES_WOBBUFFET);
+        PLAYER(SPECIES_WYNAUT) { HP(1); } ;
+        OPPONENT(SPECIES_ABRA) { HP(1); } ;
+        OPPONENT(SPECIES_KADABRA) { HP(1); } ;
```
Comment (mrgriffin) [inline test/move_effect_explosion.c:78]:
```suggestion
        PLAYER(SPECIES_WYNAUT) { HP(1); }
        OPPONENT(SPECIES_ABRA) { HP(1); }
        OPPONENT(SPECIES_KADABRA) { HP(1); }
```

File: test/move_effect_explosion.c [hunk 10]
```diff
@@ -17,12 +16,29 @@ SINGLE_BATTLE_TEST("Explosion causes the user to faint")
...
         ANIMATION(ANIM_TYPE_MOVE, MOVE_EXPLOSION, player);
+        MESSAGE("Wobbuffet fainted!");
+    }
+}
+
+SINGLE_BATTLE_TEST("Explosion causes the user & the target to faint")
+{
+    GIVEN {
+        PLAYER(SPECIES_WOBBUFFET);
+        OPPONENT(SPECIES_WOBBUFFET) { HP(1); } ;
```
Comment (mrgriffin) [inline test/move_effect_explosion.c:27]:
```suggestion
        OPPONENT(SPECIES_WOBBUFFET) { HP(1); }
```

File: test/ability_volt_absorb.c [hunk 11]
```diff
@@ -63,6 +63,30 @@ SINGLE_BATTLE_TEST("Volt Absorb is only triggered once on multi strike moves")
...
 }
 
+DOUBLE_BATTLE_TEST("Volt Absorb does not stop Electric Typed Explosion from damaging other pokemon", s16 damage1, s16 damage2) // Fixed issue #1961
+{
+    GIVEN {
+        ASSUME(gBattleMoves[MOVE_EXPLOSION].effect == EFFECT_EXPLOSION);
+        ASSUME(gBattleMoves[MOVE_EXPLOSION].type == TYPE_NORMAL);
+        PLAYER(SPECIES_JOLTEON) { Ability(ABILITY_VOLT_ABSORB); HP(1); MaxHP(TEST_MAX_HP); };
+        PLAYER(SPECIES_ABRA);
+        OPPONENT(SPECIES_GRAVELER_ALOLAN) { Ability(ABILITY_GALVANIZE); };
```
Comment (mrgriffin) [inline test/ability_volt_absorb.c:73]:
```suggestion
        PLAYER(SPECIES_JOLTEON) { Ability(ABILITY_VOLT_ABSORB); HP(1); MaxHP(TEST_MAX_HP); }
        PLAYER(SPECIES_ABRA);
        OPPONENT(SPECIES_GRAVELER_ALOLAN) { Ability(ABILITY_GALVANIZE); }
```

File: test/ability_damp.c [hunk 12]
```diff
@@ -19,6 +19,26 @@ SINGLE_BATTLE_TEST("Damp prevents explosion-like moves from enemies")
...
 
+DOUBLE_BATTLE_TEST("Damp prevents explosion-like moves from enemies in a double battle")
+{
+    u32 move;
+    PARAMETRIZE { move = MOVE_EXPLOSION; }
+    PARAMETRIZE { move = MOVE_SELF_DESTRUCT; }
+    PARAMETRIZE { move = MOVE_MIND_BLOWN; }
+    PARAMETRIZE { move = MOVE_MISTY_EXPLOSION; }
+    GIVEN {
+        PLAYER(SPECIES_PARAS) { Ability(ABILITY_DAMP); };
```
Comment (mrgriffin) [inline test/ability_damp.c:30]:
```suggestion
        PLAYER(SPECIES_PARAS) { Ability(ABILITY_DAMP); }
```

**Review by mrgriffin — DISMISSED** [review]
lgtm :+1:  all proposed changes are just small syntax tweaks

**Review by mrgriffin — APPROVED** [review]

**Pull request merged**